File: apps/driver-android/src/main/kotlin/lk/mageride/driver/support/SupportScreen.kt
package lk.mageride.driver.support

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material.icons.outlined.Search
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.driver.R
import lk.mageride.driver.ui.component.DashboardBanner
import lk.mageride.driver.ui.component.MageRideCtaTonal
import lk.mageride.driver.ui.component.SectionLabel
import lk.mageride.driver.ui.component.StatusPill
import lk.mageride.driver.ui.theme.ControlTokens
import lk.mageride.driver.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.support.FaqSummary
import lk.mageride.shared.data.models.support.Ticket
import org.koin.androidx.compose.koinViewModel

/**
 * **SCR-DA-033 · support and the daily-fee refund** (US-16.1/16.2, US-9.23).
 *
 * The wireframe, top to bottom: a `‹ Support` app bar, *"🔍 Search help"*, **Quick actions** with the
 * primary-outlined *"Request daily-fee refund … Request"* card, **Your tickets** with a status chip
 * per row, and the tonal *"Raise a ticket"* CTA.
 *
 * **Both entry points open the same sheet** (SCR-DA-033a) and both post `POST /v1/support/tickets`;
 * only `category` differs, and it is what routes the row to Finance or Support. See
 * [SupportViewModel].
 *
 * The FAQ list under the search box is US-16.1's half of the screen. The wireframe draws the search
 * field and no results, because it draws the resting state — typing is what reveals them.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun SupportScreen(onBack: () -> Unit, modifier: Modifier = Modifier) {
    val viewModel: SupportViewModel = koinViewModel()
    val state by viewModel.state.collectAsStateWithLifecycle()

    Scaffold(
        modifier = modifier.fillMaxSize(),
        topBar = {
            TopAppBar(
                title = { Text(text = stringResource(R.string.support_title)) },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Outlined.ArrowBack,
                            contentDescription = stringResource(R.string.action_back),
                        )
                    }
                },
            )
        },
    ) { insets ->
        Column(
            modifier = Modifier
                .padding(insets)
                .fillMaxSize()
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        ) {
            state.error?.let { message ->
                DashboardBanner(text = stringResource(message), accent = MaterialTheme.colorScheme.error)
            }
            state.raisedTicketId?.let {
                DashboardBanner(
                    text = stringResource(R.string.support_ticket_raised),
                    accent = MageRideTheme.status.success,
                )
            }

            Column(
                modifier = Modifier.padding(MageRideTheme.spacing.md),
                verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
            ) {
                SearchField(value = state.search, onValueChange = viewModel::onSearchChange)

                FaqResults(
                    articles = state.visibleArticles,
                    searching = state.search.isNotBlank(),
                    onOpen = viewModel::openArticle,
                )

                SectionLabel(text = stringResource(R.string.support_quick_actions))
                RefundCard(onRequest = { viewModel.openTicketSheet(SupportCategories.DAILY_FEE_REFUND) })

                SectionLabel(text = stringResource(R.string.support_your_tickets))
                TicketList(state = state, onOpen = viewModel::openTicket)

                MageRideCtaTonal(
                    label = stringResource(R.string.support_raise_ticket),
                    onClick = { viewModel.openTicketSheet(SupportCategories.GENERAL) },
                )
            }
        }
    }

    SupportSheets(state = state, viewModel = viewModel)
}

/** The wireframe's `.searchbar`. Filters what has been read — see [SupportState.visibleArticles]. */
@Composable
private fun SearchField(value: String, onValueChange: (String) -> Unit, modifier: Modifier = Modifier) {
    OutlinedTextField(
        value = value,
        onValueChange = onValueChange,
        modifier = modifier
            .fillMaxWidth()
            .height(ControlTokens.SearchBar + MageRideTheme.spacing.xs),
        placeholder = { Text(text = stringResource(R.string.support_search_hint)) },
        leadingIcon = { Icon(imageVector = Icons.Outlined.Search, contentDescription = null) },
        singleLine = true,
        shape = RoundedCornerShape(MageRideTheme.radius.md),
        textStyle = MaterialTheme.typography.bodyMedium,
    )
}

/**
 * US-16.1's articles.
 *
 * Drawn only while something is typed: the wireframe's resting state is the search field with the
 * quick actions under it, and a full FAQ index pushed above *"Request daily-fee refund"* would bury
 * the one action a driver opens this screen for.
 */
@Composable
private fun FaqResults(
    articles: List<FaqSummary>,
    searching: Boolean,
    onOpen: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    if (!searching) return

    Column(modifier = modifier.fillMaxWidth()) {
        if (articles.isEmpty()) {
            Text(
                text = stringResource(R.string.support_search_empty),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            return@Column
        }
        articles.forEach { article ->
            Text(
                text = article.title,
                modifier = Modifier
                    .fillMaxWidth()
                    .clickable { onOpen(article.articleId) }
                    .padding(vertical = MageRideTheme.spacing.xs),
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurface,
            )
        }
    }
}

/**
 * *"Request daily-fee refund · e.g. app crashed on Go Online"* (US-9.23, US-14.11).
 *
 * The wireframe outlines this card in `primary` — it is the one action on the screen that moves
 * money. **Request** opens SCR-DA-033a with the category fixed, because a refund is a ticket with a
 * category and not an endpoint of its own; the reversal itself is an admin decision (US-14.11).
 */
@Composable
private fun RefundCard(onRequest: () -> Unit, modifier: Modifier = Modifier) {
    Card(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(MageRideTheme.radius.md),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface),
        border = BorderStroke(ControlTokens.Border, MaterialTheme.colorScheme.primary),
    ) {
        Row(
            modifier = Modifier.padding(MageRideTheme.spacing.sm),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = stringResource(R.string.support_refund_title),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurface,
                )
                Text(
                    text = stringResource(R.string.support_refund_hint),
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            TextButton(onClick = onRequest) {
                Text(text = stringResource(R.string.support_refund_action))
            }
        }
    }
}

/** *"Your tickets"* — `#TK-9012 · Fee charged in error` with its status chip. */
@Composable
private fun TicketList(state: SupportState, onOpen: (String) -> Unit, modifier: Modifier = Modifier) {
    Column(modifier = modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs)) {
        when {
            state.loading -> Unit

            state.tickets.isEmpty() -> Text(
                text = stringResource(R.string.support_tickets_empty),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )

            else -> state.tickets.forEach { ticket ->
                TicketRow(ticket = ticket, onClick = { onOpen(ticket.ticketId) })
            }
        }
    }
}

/**
 * One ticket row.
 *
 * The wireframe prints `#TK-9012`; a ticket id on this platform is a ULID and there is no `TK-`
 * series to render — the same finding C074 recorded about `DRV-22011`. The row therefore leads with
 * the **category**, which is what the driver recognises, and the id is not drawn at all: a
 * twenty-six-character identifier in a list row is noise nobody reads out.
 */
@Composable
private fun TicketRow(ticket: Ticket, onClick: () -> Unit, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(vertical = MageRideTheme.spacing.xs),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        Text(
            text = categoryLabel(ticket.category),
            modifier = Modifier.weight(1f),
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurface,
        )
        StatusPill(
            label = stringResource(SupportLabels.status(ticket.status)),
            tone = SupportLabels.tone(ticket.status),
        )
    }
}

File: apps/driver-ios/DriverApp/Support/SupportScreen.swift
import MageRideShared
import SwiftUI

/// **SCR-DI-033 · support and the daily-fee refund** (US-16.1/16.2, US-9.23).
///
/// The wireframe, top to bottom: a `Support` large title, *"🔍 Search help"*, **Quick actions** with
/// the primary-outlined *"Request daily-fee refund … Request"* card, **Your tickets** with a status
/// chip per row, and the tonal *"Raise a ticket"* CTA.
///
/// **Both entry points open the same sheet** (SCR-DI-033a) and both post `POST /v1/support/tickets`;
/// only `category` differs, and it is what routes the row to Finance or Support. See
/// ``SupportModel``.
///
/// The FAQ list under the search box is US-16.1's half of the screen. The wireframe draws the search
/// field and no results, because it draws the resting state — typing is what reveals them.
///
/// It hangs off the **Menu** tab (``DriverRoute/tab``), which is what makes the system back button
/// say `‹ Menu` rather than the Android twin's own `‹ Support` app bar.
@MainActor
struct SupportScreen: View {

    @StateObject private var model: SupportModel

    init(model: @autoclosure @escaping () -> SupportModel) {
        _model = StateObject(wrappedValue: model())
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                if let errorKey = model.state.errorKey {
                    DashboardBanner(text: errorKey.localised, accent: MageRideColor.error)
                        .onTapGesture(perform: model.dismissError)
                }
                if model.state.raisedTicketId != nil {
                    DashboardBanner(
                        text: "support_ticket_raised".localised,
                        accent: MageRideColor.success,
                        symbolName: "checkmark.circle.fill"
                    )
                }

                SearchField(
                    placeholderKey: "support_search_hint",
                    value: Binding(get: { model.state.search }, set: model.onSearchChange)
                )

                if model.state.isSearching {
                    faqResults
                }

                SectionLabel(key: "support_quick_actions")
                refundCard

                SectionLabel(key: "support_your_tickets")
                ticketList

                Button { model.openTicketSheet(category: SupportCategories.general) } label: {
                    Text(key: "support_raise_ticket")
                }
                .buttonStyle(.mageCtaTonal)
                .padding(.top, MageRideSpacing.xxs)
            }
            .padding(MageRideSpacing.md)
        }
        .background(MageRideColor.surface)
        .navigationTitle(Text(key: "support_title"))
        .navigationBarTitleDisplayMode(.large)
        .task { await model.refresh() }
        // One binding for three overlays. SwiftUI presents one sheet per context and silently drops
        // the rest, which is the trap C091 recorded on SCR-DI-022.
        .sheet(item: Binding(get: { model.state.sheet }, set: { if $0 == nil { model.closeSheet() } })) { sheet in
            SupportSheets(sheet: sheet, model: model)
        }
    }

    /// US-16.1's articles, drawn only while something is typed.
    @ViewBuilder
    private var faqResults: some View {
        if model.state.visibleArticles.isEmpty {
            Text(key: "support_search_empty")
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
        } else {
            // Keyed on the article rather than on an index pair: **Swift has no key paths into
            // tuples** (the C087 finding), so `Array(_:enumerated())` cannot supply a `ForEach` id
            // here or anywhere else in this cluster. "Is this the last row" is asked of the
            // collection instead.
            let articles = model.state.visibleArticles
            GroupedList {
                ForEach(articles, id: \.articleId) { article in
                    Button {
                        Task { await model.openArticle(articleId: article.articleId) }
                    } label: {
                        HStack(spacing: MageRideSpacing.xs) {
                            Text(article.title)
                                .mageFont(.body)
                                .foregroundStyle(MageRideColor.onSurface)
                                .multilineTextAlignment(.leading)
                            Spacer(minLength: MageRideSpacing.xs)
                            Image(systemName: "chevron.right")
                                .font(.caption)
                                .foregroundStyle(MageRideColor.outlineVariant)
                        }
                        .padding(.horizontal, MageRideSpacing.sm)
                        .frame(minHeight: MageRideControl.minimumTapTarget)
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)

                    if article.articleId != articles.last?.articleId {
                        Rectangle()
                            .fill(MageRideColor.surfaceVariant)
                            .frame(height: MageRideControl.hairline)
                            .padding(.leading, MageRideSpacing.sm)
                    }
                }
            }
        }
    }

    /// *"Request daily-fee refund · e.g. app crashed on Go Online"* (US-9.23, US-14.11).
    ///
    /// The wireframe outlines this card in `primary` — it is the one action on the screen that moves
    /// money. **Request** opens SCR-DI-033a with the category fixed, because a refund is a ticket
    /// with a category and not an endpoint of its own; the reversal itself is an admin decision.
    private var refundCard: some View {
        HStack(spacing: MageRideSpacing.xs) {
            VStack(alignment: .leading, spacing: 1) {
                Text(key: "support_refund_title")
                    .mageFont(.body)
                    .foregroundStyle(MageRideColor.onSurface)
                Text(key: "support_refund_hint")
                    .mageFont(.caption)
                    .foregroundStyle(MageRideColor.onSurfaceVariant)
            }
            Spacer(minLength: MageRideSpacing.xs)

            Button { model.openTicketSheet(category: SupportCategories.dailyFeeRefund) } label: {
                Text(key: "support_refund_action")
                    .mageFont(.subtitle)
                    .foregroundStyle(MageRideColor.primary)
                    .frame(minHeight: MageRideControl.minimumTapTarget)
            }
            .buttonStyle(.plain)
        }
        .padding(MageRideSpacing.sm)
        .frame(maxWidth: .infinity)
        .background(
            MageRideColor.background,
            in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
        )
        .overlay {
            RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
                .strokeBorder(MageRideColor.primary, lineWidth: 2)
        }
    }

    /// *"Your tickets"* — the category with its status chip.
    @ViewBuilder
    private var ticketList: some View {
        if model.state.isLoading {
            ProgressView()
                .frame(maxWidth: .infinity)
        } else if model.state.tickets.isEmpty {
            Text(key: "support_tickets_empty")
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
        } else {
            GroupedList {
                ForEach(model.state.tickets, id: \.ticketId) { ticket in
                    ticketRow(ticket, showsSeparator: ticket.ticketId != model.state.tickets.last?.ticketId)
                }
            }
        }
    }

    /// One ticket row.
    ///
    /// The wireframe prints `#TK-9012`; a ticket id on this platform is a ULID and there is no `TK-`
    /// series to render — the same finding C092 recorded about `DRV-22011`. The row therefore leads
    /// with the **category**, which is what the driver recognises, and the id is not drawn at all: a
    /// twenty-six-character identifier in a list row is noise nobody reads out.
    private func ticketRow(_ ticket: Ticket, showsSeparator: Bool) -> some View {
        VStack(spacing: 0) {
            Button {
                Task { await model.openTicket(ticketId: ticket.ticketId) }
            } label: {
                HStack(spacing: MageRideSpacing.xs) {
                    Text(SupportLabels.category(ticket.category))
                        .mageFont(.body)
                        .foregroundStyle(MageRideColor.onSurface)
                        .multilineTextAlignment(.leading)
                    Spacer(minLength: MageRideSpacing.xs)
                    StatusPill(
                        label: SupportLabels.statusKey(ticket.status).localised,
                        tone: SupportLabels.tone(ticket.status)
                    )
                }
                .padding(.horizontal, MageRideSpacing.sm)
                .frame(minHeight: MageRideControl.minimumTapTarget)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityElement(children: .combine)

            if showsSeparator {
                Rectangle()
                    .fill(MageRideColor.surfaceVariant)
                    .frame(height: MageRideControl.hairline)
                    .padding(.leading, MageRideSpacing.sm)
            }
        }
    }
}
